src/main/java/org/openstreetmap/josm/plugins/pt_assistant/routeexplorer/transportmode/AbstractTransportMode.java

```diff
@@ -0,0 +1,98 @@
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import com.drew.lang.annotations.NotNull;
+import org.openstreetmap.josm.data.osm.IRelation;
+import org.openstreetmap.josm.data.osm.IWay;
+import org.openstreetmap.josm.data.osm.Node;
+import org.openstreetmap.josm.data.osm.OsmPrimitive;
+import org.openstreetmap.josm.data.osm.OsmPrimitiveType;
+import org.openstreetmap.josm.data.osm.Relation;
+import org.openstreetmap.josm.data.osm.Way;
+import org.openstreetmap.josm.plugins.pt_assistant.routeexplorer.WayTraversalDirection;
+import org.openstreetmap.josm.plugins.pt_assistant.utils.WayUtils;
+
+public abstract class AbstractTransportMode implements ITransportMode {
+    String modeOfTransport = "";
+    String[] additionalTypesForTurnRestriction;
+    String[] turnRestrictionExceptionsFor;
+    String[] oneWayExceptionsFor;
+
+    @Override
+    public boolean canBeUsedForRelation(@NotNull final IRelation<?> relation) {
+        return relation.hasTag("type", "route");
+    }
+
+    @Override
+    public boolean canTraverseWay(@NotNull final IWay<?> way, @NotNull final WayTraversalDirection direction) {
+        final String oneway = way.get("oneway");
+        return oneway == null
+                || Arrays.stream(oneWayExceptionsFor)
+                    .map(mode -> way.get("oneway:" + mode))
+                    .anyMatch("no"::equals)
+                || ("yes".equals(oneway) && direction == WayTraversalDirection.FORWARD)
+                || ("-1".equals(oneway) && direction == WayTraversalDirection.BACKWARD);
+    }
+
+    @Override
+    public boolean canTurn(@NotNull final Way from, @NotNull final Node via, @NotNull final Way to) {
+        List<String> types = new java.util.ArrayList<>(Collections.singletonList("restriction"));
+        Arrays.stream(additionalTypesForTurnRestriction).map(at -> "restriction:" + at).forEach(types::add);
+        final boolean bothWaysAreConnectedAtViaNode = from.containsNode(via) && to.containsNode(via);
+        for (Relation restrictionRelation : getTurnRestrictions(from, via, to, types)) {
+            for (String type : types) {
+                final String restriction = restrictionRelation.get(type);
+                if (restriction != null && restriction.startsWith("no_")) {
+                    final String except = !"".equals(modeOfTransport) ? restrictionRelation.get("except") : "";
+                    if (except != null) {
+                        for (String tre : turnRestrictionExceptionsFor) {
+                            if (except.contains(tre)) {
+                                return bothWaysAreConnectedAtViaNode;
+                            }
+                        }
+                    }
+                    return false;
+                }
+            }
+        }
+        return bothWaysAreConnectedAtViaNode;
+    }
+
+    @Override
+    public boolean canTurn(@NotNull final Way from, @NotNull final Way via, @NotNull final Way to) {
```

Comment:
You cannot check this that easily. Have a look at the [specs](https://wiki.openstreetmap.org/wiki/Relation:restriction#Members): Via can be a list of multiple ways.

The problem with turn restrictions is, that you cannot step through the route easilly, because you never know if they are part of a via or not. 

if have not seen any usages of canTurn. What is the use case of it?

Reply:
… and this is my suggestion to check turn restrictions:

* Inputs
** ``List<PartOfWay> route` All previous ways traversed as list (more than we need, but they are for free)
** `PartOfWay next` The next way we want to add
* Returns
** true iff turning into that way is not blocked by any turn relation. (=> Returns true even if we might traverse the next segment but would be blocked later by a no_*** relation)
* Algorithm
** (Check for no_* restrictions)
** Let R be the set of relations for which these conditions are met
*** relation has `next` as member
*** role of `next` is 'to'
*** relation is a turn for the current transport mode with no_ type
** For each item r in R, do:
*** Check if r matches current state. If it does => return false;
** (check for only_* relations)
** Let last be the last way of ways.
** Let O be the set of relations with:
*** Relation is a only_* relation for the current transport type
*** last is a `from` or `via` member of the relation.
*** If it is a `via` member, the previous members need to match the `route`
** If O is empty, return true (no only_candidates)
** If any relation in O has `next` as next way (=> the way after `last`, no matter if it was a via or to), return true
** In all other cases return false

*To check if no_* turn relation matches*
* If the relation has via ways:
** Traverse the `route` backwards and check if it matches the via ways
** Now traverse back one more way and check for the `from` way.
* If the relation has via nodes:
** Check if the via node is the connecting node.

Reply:
When the routing helper/ route explorer arrives at a gap, it would continue until the next junction.

At the junction it needs to propose ways that are

1. suitable for the mode of transport
2. the mode of transport is allowed to turn onto

We are creating the building blocks that can be used later on, so it's to be expected that they are not used yet. But they will be.

Thanks for the suggested workflow. I'll try to implement it. It is a lot more complex than I had expected it would be. But maybe that was to be expected :-)